src/core/ddsi/src/q_entity.c

```diff
@@ -5570,6 +5799,236 @@ uint64_t get_entity_instance_id (const struct ddsi_domaingv *gv, const struct dd
+  {
+    tpd = new_topic_definition (gv, type, qos);
+    assert (tpd != NULL);
+    new = true;
+  }
+  if (new_tpd != NULL)
+    *new_tpd = new;
+  ddsrt_mutex_unlock (&gv->topic_defs_lock);
+  return tpd;
+}
+
+dds_return_t lookup_topic_definition_by_name (struct ddsi_domaingv *gv, const char * topic_name, struct topic_definition **topic_definition)
+{
+  assert (topic_definition != NULL);
+  *topic_definition = NULL;
+  struct ddsrt_hh_iter it;
+  dds_return_t ret = DDS_RETCODE_OK;
+  ddsrt_mutex_lock (&gv->topic_defs_lock);
+  for (struct topic_definition *tpd = ddsrt_hh_iter_first (gv->topic_defs, &it); tpd; tpd = ddsrt_hh_iter_next (&it))
+  {
+    if (!strcmp (tpd->xqos->topic_name, topic_name))
+    {
+      if (*topic_definition == NULL)
+        *topic_definition = tpd;
+      else
+      {
+        *topic_definition = NULL;
+        ret = DDS_RETCODE_PRECONDITION_NOT_MET;
+        break;
+      }
+    }
+  }
+  ddsrt_mutex_unlock (&gv->topic_defs_lock);
+  return ret;
+}
+
+static void gc_delete_topic_definition (struct gcreq *gcreq)
+{
+  struct topic_definition *tpd = gcreq->arg;
+  struct ddsi_domaingv *gv = tpd->gv;
+  GVLOGDISC ("gcreq_delete_topic_definition(%p)\n", (void *) gcreq);
+  gcreq_free (gcreq);
+  if (tpd->type != NULL)
+    ddsi_sertype_unref ((struct ddsi_sertype *) tpd->type);
+  ddsi_xqos_fini (tpd->xqos);
+  ddsrt_free (tpd->xqos);
+  ddsrt_free (tpd);
+}
+
+static int delete_topic_definition (struct topic_definition *tpd, ddsrt_wctime_t timestamp)
+{
+  struct ddsi_domaingv *gv = tpd->gv;
+  GVLOGDISC ("delete_topic_definition (%p) ", tpd);
+  ddsrt_mutex_lock (&gv->topic_defs_lock);
+#ifndef NDEBUG
+  void *tmp = ddsrt_hh_lookup (gv->topic_defs, tpd);
+  assert (tmp != NULL);
+#endif
+  (void) ddsrt_hh_remove (gv->topic_defs, tpd);
+  ddsrt_mutex_unlock (&gv->topic_defs_lock);
+  builtintopic_write_topic (gv->builtin_topic_interface, tpd, timestamp, false);
+  GVLOGDISC ("- deleting\n");
+  gcreq_topic_definition (tpd);
+  return DDS_RETCODE_OK;
+}
+
+
+/* PROXY-TOPIC --------------------------------------------------- */
+
+int proxy_topic_equal (const struct proxy_topic *proxy_tp_a, const struct proxy_topic *proxy_tp_b)
+{
+  if (proxy_tp_a != NULL && proxy_tp_b != NULL)
+    return topic_definition_equal (proxy_tp_a->definition, proxy_tp_b->definition);
+  return proxy_tp_a == proxy_tp_b;
+}
+
+bool new_proxy_topic (struct proxy_participant *proxypp, const ddsi_guid_t *guid, const type_identifier_t *type_id, struct dds_qos *qos, ddsrt_wctime_t timestamp)
+{
+  assert (proxypp != NULL);
+  struct ddsi_domaingv *gv = proxypp->e.gv;
+  bool new_tpd = false, found_proxytp = false;
+  proxy_topic_list_iter_t it;
+  struct topic_definition *tpd = lookup_topic_definition (gv, qos, type_id, NULL, &new_tpd);
+  ddsrt_mutex_lock (&proxypp->e.lock);
+  for (struct proxy_topic *proxytp = proxy_topic_list_iter_first (&proxypp->topics, &it); proxytp != NULL && !found_proxytp; proxytp = proxy_topic_list_iter_next (&it))
```

Comment:
I am not entirely happy with all the linear scans to find topics (by name in`lookup_topic_definition_by_name` , by definition here, in `delete_proxy_topic`, perhaps also elsewhere): before you know it all starts to add up to real work.

That said, I don't want to derail this PR by asking for a rewrite of all this. It is optional (and excluded by default) so we can accept a suboptimal implementation for the time being.